### ISSUE TITLE
[BUG] Missing character when running inference on a big number of prompts

### ISSUE THREAD

**Issue by drneverend** [ModelTC/lightllm#153]
**Before you submit an issue, please search for existing issues to avoid duplicates.**
Yes.

**Issue description:**
When running inference for a big number of prompts (like 5000 and more) with llama2-7b model, some of the characters in the results might be missing. We reproduced it on both V100 and MI200.

max sequence length: 2048
max new tokens: 256

An example of missing character issue.
Expected result: 企业微信
Actual result: 企业信

**Environment:**

- OS: (Ubuntu 20.04
- GPU info:
  - V100, CUDA 11.8
- Python: 3.9
- LightLLm: 2d3cd3330ed7526864bb8c17da4a4bbec6aa29c7
- openai-triton: the version specified in requirement.txt

**Comment by hiworldwzj** [comment]
@drneverend  Hello. 这个是已知的情况，在正常推理的时候有时候数值波动可能会带来一些小的输出波动，这个在过去的运行过程中也遇到过一些，但是一般输出的句子还是通顺的。
这个具体的原因目前还没有被找到，只是怀疑是不是输出的概率 ”微“ 和 ”信" 两个token的概率特别相近，一些细微的波动就造成了选择了不同的结果。

**Comment by garycaokai** [comment]
我也遇到这个问题，llama结构，超长query（> 8000token）有时候会出现掉字/多字的情况，相同query，vllm返回结果正常，
输入：human:xxx\n问题:2022年年度，公司实现销售回款是多少亿?\nsystem:
输出：
{
    "generated_text": [
        "20222年年度，公司实现销售回款48亿元。"
    ]
}

**Comment by hiworldwzj** [comment]
@garycaokai 这个超长context场景，可能是使用的位置向量机制没有正确的加载一些参数，现在的加载位置向量有些hard code，请问你使用的是什么模型，我看看是不是没有适配好位置向量加载的部分。

**Comment by hiworldwzj** [comment]
@garycaokai 也有可能是tokenizer的处理带来的问题，可以给一些更详细的信息吗，比如模型，输入，请求采样参数等，方便debug。

**Comment by garycaokai** [comment]
类似llama的模型，看实现和vllm也差不多，初步怀疑是float16精度不够
lightllm/models/llama/model.py
        inv_freq = 1.0 / (base ** (torch.arange(0, self.head_dim_, 2, device="cpu", dtype=torch.float32) / self.head_dim_))
        t = torch.arange(max_seq_len + 1024 * 64, device="cpu", dtype=torch.float32) / rope_scaling_factor
        freqs = torch.outer(t, inv_freq)

        self._cos_cached = torch.cos(freqs).to(torch.float16).cuda()
        self._sin_cached = torch.sin(freqs).to(torch.float16).cuda()
 ==========
 vllm(vllm/model_executor/layers/attention.py)
       nv_freq = 1.0 / (base**(torch.arange(
            0, rotary_dim, 2, dtype=torch.float, device="cuda") / rotary_dim))
        t = torch.arange(max_position, dtype=torch.float, device="cuda")
        freqs = torch.einsum("i,j -> ij", t, inv_freq)
        cos = freqs.cos()
        sin = freqs.sin()
        cache = torch.cat((cos, sin), dim=-1)

        # FIXME(woosuk): This assumes that we configure the default dtype when
        # initializing the model.
        # TODO(woosuk): Make it more robust.
        torch_dtype = torch.get_default_dtype()
        cache = cache.to(torch_dtype)
        # Embedding size: [max_position, rotary_dim]
        self.register_buffer("cos_sin_cache", cache, persistent=False)

**Comment by shangshng** [comment]
同样的问题，用的模型是llama2 7b/13b，给的prompt token长度在2100+~2600+，prompt数量在5000条左右，前一两百条漏字很严重，后面的还行。运行代码直接在benchmark上改的，依然使用request rate=200
用的sample参数：
`"parameters": {
    "do_sample": False,
    "ignore_eos": False,
    "max_new_tokens": 256,
    "temperature": 0.0,
    "top_p": 1.0,
    "top_k": -1}`

**Comment by garycaokai** [comment]
可能是triton版本问题，改用了最新的lightllm镜像，triton环境成了2.1.0，缺字问题没有复现

> 类似llama的模型，看实现和vllm也差不多，初步怀疑是float16精度不够 lightllm/models/llama/model.py inv_freq = 1.0 / (base ** (torch.arange(0, self.head_dim_, 2, device="cpu", dtype=torch.float32) / self.head_dim_)) t = torch.arange(max_seq_len + 1024 * 64, device="cpu", dtype=torch.float32) / rope_scaling_factor freqs = torch.outer(t, inv_freq)
> 
> ```
>     self._cos_cached = torch.cos(freqs).to(torch.float16).cuda()
>     self._sin_cached = torch.sin(freqs).to(torch.float16).cuda()
> ```
> 
> ========== vllm(vllm/model_executor/layers/attention.py) nv_freq = 1.0 / (base**(torch.arange( 0, rotary_dim, 2, dtype=torch.float, device="cuda") / rotary_dim)) t = torch.arange(max_position, dtype=torch.float, device="cuda") freqs = torch.einsum("i,j -> ij", t, inv_freq) cos = freqs.cos() sin = freqs.sin() cache = torch.cat((cos, sin), dim=-1)
> 
> ```
>     # FIXME(woosuk): This assumes that we configure the default dtype when
>     # initializing the model.
>     # TODO(woosuk): Make it more robust.
>     torch_dtype = torch.get_default_dtype()
>     cache = cache.to(torch_dtype)
>     # Embedding size: [max_position, rotary_dim]
>     self.register_buffer("cos_sin_cache", cache, persistent=False)
> ```

j

**Comment by shangshng** [comment]
这里debug了一下，问题应该是出在httpserver的generate和handle_loop的event同步上

>        
          while True:
              try:
                  await asyncio.wait_for(event.wait(), timeout=5) #这里timeout改为None就没有丢字现象了
              except asyncio.TimeoutError:
                  pass
              event.clear()
              out_str, metadata, finished, _ = self.req_id_to_out_inf[request_id]
              if len(metadata) != 0:
                  self.req_id_to_out_inf[request_id] = ("", {}, finished, event)
                  metadata["prompt_tokens"] = prompt_tokens
                  yield out_str, metadata, finished
              if finished:
                  try:
                      del self.req_id_to_out_inf[request_id]
                  except:
                      pass
                  break
但是这么改了后throughput会降低，请问 @hiworldwzj 对这个有什么想法吗？

**Comment by hiworldwzj** [comment]
@shangshng 感谢， 最近在 #216 修复了这个问题。
